### PR TITLE
1099 Add support to /oauth/revoke endpoint.

### LINKED ISSUE #1099: Support /oauth/revoke endpoint
This issue is to support the new application /oauth/revoke endpoint that implements the revocation functionality as can been in the [Oauth Spec](https://tools.ietf.org/html/rfc7009
)

### PR DESCRIPTION
**Pull request by josebarrueta**
fixes #1099

### CODE REVIEW

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy [hunk 1]
```diff
@@ -2149,6 +2156,192 @@ class ApplicationIT extends ClientIT {
...
+            accessTokens.add(accessToken)
+
+            refreshToken = refreshToken == null ? refreshTokenResult.refreshToken : refreshToken
+        }
+
+        AccessToken toRevoke = accessTokens.get(3)
+
+        OAuthTokenRevocators.OAUTH_TOKEN_REVOCATOR.forApplication(app).revoke(OAuthRequests.OAUTH_TOKEN_REVOCATION_REQUEST.builder().setToken(toRevoke.getJwt()).build())
+
+        accessTokens.each {
```
Comment (dogeared) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy]: If you are revoking a single access token from the list, why should they all be revoked?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java [hunk 2]
```diff
@@ -170,7 +169,7 @@ public void init() {
         Assert.notNull(resultFactory, "accessTokenResultFactory cannot be null.");
         Assert.notNull(accountSaver, "accountSaver cannot be null.");
         Assert.notNull(eventPublisher, "eventPublisher cannot be null.");
-        Assert.notNull(grantTypeStatusValidator, "grantTypeStatusValidator cannot be null.");
+        Assert.notNull(grantTypeValidator, "grantTypeStatusValidator cannot be null.");
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java]: message should be: `grantTypeValidator cannot be null.`

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java [hunk 3]
```diff
@@ -356,7 +355,7 @@ protected ViewModel doPost(HttpServletRequest request, HttpServletResponse respo
 
         } catch (OAuthException e) {
 
-            log.debug("OAuth Access Token request failed. Root cause: " + e.getMessage());
+            log.debug("OAuth Access Token request failed. Root cause: {}", e.getMessage());
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java]: nice catch!

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java [hunk 4]
```diff
@@ -389,9 +418,46 @@ private void uncache(String cacheKey, Class<? extends Resource> resourceType) {
...
+            String nonSignedToken = token.substring(0, signatureIndex + 1);
+            Jwt<Header, Claims> jwt = Jwts.parser().parseClaimsJwt(nonSignedToken);
+
+            Object stt = jwt.getHeader().get("stt");
+
+            String rti = null;
+
+            Claims body = jwt.getBody();
+
+            if ("refresh".equals(stt) || "refresh_token".equals(tokenTypeHint)) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: Why do we need hint and `stt`? Are there situations where we have a token without `stt`? If not, then why do we need tokenTypeHint at all?

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java [hunk 5]
```diff
@@ -389,9 +418,46 @@ private void uncache(String cacheKey, Class<? extends Resource> resourceType) {
...
     }
 
-    private boolean isApiKeyCollectionQuery(ResourceDataRequest request) {
-        return ApiKeyList.class.isAssignableFrom(request.getResourceClass()) &&
-                request.getUri().hasQuery() && request.getUri().getQuery().containsKey(ID.getName());
+    private void uncacheRevokedToken(Map<String, Object> data) {
+
+        String token = data.get(OAuthTokenRevocationAttempt.TOKEN.getName()).toString();
+
+        int signatureIndex = token.lastIndexOf('.');
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: This seems hacky to me. If it's an unsigned token, in our use cases, then there's something wrong, no?

File: impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java [hunk 6]
```diff
@@ -452,6 +453,8 @@ public ResourceDataResult filter(final ResourceDataRequest req) {
                     // Fix for https://github.com/stormpath/stormpath-sdk-java/issues/218
                     if ( response.getHttpStatus() == 202 ) { //202 means that the request has been accepted for processing, but the processing has not been completed. Therefore we do not have a response body.
                         responseBody = java.util.Collections.emptyMap();
+                    } else if(response.getHttpStatus() == 200 && OAuthTokenRevoked.class.isAssignableFrom(returnType)) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java]: won't the response be 404 for a revoked token?

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy (hunk 1 above)
Comment (josebarrueta) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy]: propagation.

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java (hunk 5 above)
Comment (josebarrueta) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: Worst side effect is removing something from cache, so I don't think it's worth it to do more validations.

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java (hunk 4 above)
Comment (josebarrueta) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: The developer might remove the stt header from the token, then re-package it and send it.

File: impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java (hunk 6 above)
Comment (josebarrueta) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java]: Invalid tokens according to the spec should return 200 as well.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java (hunk 2 above)
Comment (josebarrueta) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/mvc/AccessTokenController.java]: cool

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java (hunk 4 above)
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: We use the `stt` header param to identify token type. removing it invalidates the token, I think. If users want to repackage, I think they should include the header as that's part of what contributes to it being a valid token.

Comment (josebarrueta) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: I actually was able to assert this in the last three lines of: ApplicationIT.testInvalidTokens_NoTokenType(..)

File: api/src/main/java/com/stormpath/sdk/oauth/TokenTypeHint.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    TokenTypeHint(String value) {
+        this.value = value;
+    }
+
+    public String getValue() {
+        return value;
+    }
+
+    public static TokenTypeHint fromValue(String value) {
```
Comment (josebarrueta) [inline api/src/main/java/com/stormpath/sdk/oauth/TokenTypeHint.java]: aren't the two approaches equivalent? any advantages from this one?

**Pull request merged**